internal/worker_versioning_rules.go

```diff
@@ -0,0 +1,338 @@
+				RuleIndex: v.RuleIndex,
+				Rule:      versioningAssignmentRuleToProto(&v.Rule),
+				Force:     v.Force,
+			},
+		}
+	case *VersioningOpDeleteAssignmentRule:
+		req.Operation = &workflowservice.UpdateWorkerVersioningRulesRequest_DeleteAssignmentRule{
+			DeleteAssignmentRule: &workflowservice.UpdateWorkerVersioningRulesRequest_DeleteBuildIdAssignmentRule{
+				RuleIndex: v.RuleIndex,
+				Force:     v.Force,
+			},
+		}
+	case *VersioningOpInsertRedirectRule:
+		req.Operation = &workflowservice.UpdateWorkerVersioningRulesRequest_InsertCompatibleRedirectRule{
+			InsertCompatibleRedirectRule: &workflowservice.UpdateWorkerVersioningRulesRequest_AddCompatibleBuildIdRedirectRule{
+				Rule: versioningRedirectRuleToProto(&v.Rule),
+			},
+		}
+	case *VersioningOpReplaceRedirectRule:
+		req.Operation = &workflowservice.UpdateWorkerVersioningRulesRequest_ReplaceCompatibleRedirectRule{
+			ReplaceCompatibleRedirectRule: &workflowservice.UpdateWorkerVersioningRulesRequest_ReplaceCompatibleBuildIdRedirectRule{
+				Rule: versioningRedirectRuleToProto(&v.Rule),
+			},
+		}
+	case *VersioningOpDeleteRedirectRule:
+		req.Operation = &workflowservice.UpdateWorkerVersioningRulesRequest_DeleteCompatibleRedirectRule{
+			DeleteCompatibleRedirectRule: &workflowservice.UpdateWorkerVersioningRulesRequest_DeleteCompatibleBuildIdRedirectRule{
+				SourceBuildId: v.SourceBuildId,
+			},
+		}
+	case *VersioningOpCommitBuildId:
+		req.Operation = &workflowservice.UpdateWorkerVersioningRulesRequest_CommitBuildId_{
+			CommitBuildId: &workflowservice.UpdateWorkerVersioningRulesRequest_CommitBuildId{
+				TargetBuildId: v.TargetBuildId,
+				Force:         v.Force,
+			},
+		}
+	}
+
+	return req, nil
+}
+
+type GetWorkerVersioningOptions struct {
+	// The task queue to get the versioning rules from.
+	TaskQueue string
+}
+
+func (gw *GetWorkerVersioningOptions) validateAndConvertToProto(namespace string) (*workflowservice.GetWorkerVersioningRulesRequest, error) {
+	if namespace == "" {
+		return nil, errors.New("missing namespace argument")
+	}
+
+	if gw.TaskQueue == "" {
+		return nil, errors.New("missing  TaskQueue field")
+	}
+	req := &workflowservice.GetWorkerVersioningRulesRequest{
+		Namespace: namespace,
+		TaskQueue: gw.TaskQueue,
+	}
+
+	return req, nil
+}
+
+type WorkerVersioningRules struct {
+	AssignmentRules []*VersioningAssignmentRule
+	RedirectRules   []*VersioningRedirectRule
+	ConflictToken   VersioningConflictToken
+}
+
+func versioningAssignmentRuleToProto(rule *VersioningAssignmentRule) *taskqueuepb.BuildIdAssignmentRule {
+	// Assumed `rule` already validated
+	result := &taskqueuepb.BuildIdAssignmentRule{
+		TargetBuildId: rule.TargetBuildId,
+	}
+
+	switch r := rule.Ramp.(type) {
+	case *VersioningRampByPercentage:
+		result.Ramp = &taskqueuepb.BuildIdAssignmentRule_PercentageRamp{
+			PercentageRamp: &taskqueuepb.RampByPercentage{
+				RampPercentage: r.Percentage,
+			},
+		}
+	}
+	// Ignore `rule.Timestamp`
```

Comment:
Can you clarify why timestamp is ignored? What's its purpose if unused? Should we just remove the field?

Reply:
The protos sometimes return a `TimestampedBuildIdAssignmentRule` or `*RedirectRule` that just add the timestamp to `BuildIdAssignmentRule`. Rather than expose both types I merged then making the Timestamp optional, but then I have to convert to the actual inner proto ignoring the timestamp. 
I should improve that comment...

Reply:
We have found in the past, especially with more complicated model sets, that staying close to the API structure is safer for future changes. I think our objects in Go should very closely (exactly?) align with those in the API if we can.

Reply:
+1 staying closer to the API structure makes future changes potentially much simpler

Reply:
OK, I'll make visible types TimestampedBuildIdAssignmentRule and TimestampedCompatibleBuildIdRedirectRule in the SDK, but frankly, this looks like a complete overkill to me...

Reply:
> this looks like a complete overkill to me

That's ok, the Go implementations often do